smcore/languages.h

```diff
@@ -0,0 +1,85 @@
+//******************************************************************************
+// Copyright (C) 1999 Jim Wanner and the SourceMonitor team.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a
+// copy of this software and associated documentation files (the "Software"),
+// to deal in the Software without restriction, including without limitation
+// the rights to use, copy, modify, merge, publish, distribute, sublicense,
+// and/or sell copies of the Software, and to permit persons to whom the
+// Software is furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included
+// in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+// OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+// THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+// FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+//******************************************************************************
+
+#pragma once
+
+#include <stdbool.h>
+#include "smstring.h"
+
+namespace smos
+{
+    namespace smcore
+    {
+        /**
+         * @brief Array of language strings, order equal to the enum here
+         */
+        extern const char *langguagestrings[];
+
+        class Languages
+        {
+        public:
+            /**
+             * @brief Definition of enums according to the strings in the body
+             *
+             */
+            enum class Type
+            {
+                SMC = 0,
+                SMCPP = 1,
+                SMCS = 2,
+                SMDELPHI = 3,
+                SMHTML = 4,
+                SMJAVA = 5,
+                SMVBNET = 6,
+                SMVISUALBASIC = 7
+            };
+
+            /**
+             * @brief Convert the enum to a char* as in the array languagestrings
+             * @see languagestrings
+             *
+             * @param smos::smcore::Languages::Type Kind of language to get string for
+             * @return const char* Language enum as string
+             */
+            static const char *getLanguageString(smos::smcore::Languages::Type lang);
```

Comment:
The magic strings are caused by a MFC macro (I'm not sure, but it might be DECLARE_SERIAL). While serializing these classes, the class name is written as string to the stream. Up to now I don't have any better solution. Do you know one? I'm open for a better solution.